### PR TITLE
RFC56: annotation and mutation status filtering for Study View and enrichment/count endpoints

### PR DESCRIPTION
**Pull request by pvannierop**
# Implement annotation and mutation status filtering for Study View and enrichment/count endpoints

:warning: ~This PR should be merged together with the corresponding frontend branch.~

This PR is covered by RFC56 sections [Annotation menu in Study View](https://docs.google.com/document/d/1TLRIOurdRK1HE7nDPqk6d4zhfjpgchBd9tU6PQpbwxw/edit#heading=h.ttnh4ne6gi59) and [Annotation menu in Group Comparison view](https://docs.google.com/document/d/1TLRIOurdRK1HE7nDPqk6d4zhfjpgchBd9tU6PQpbwxw/edit#heading=h.fen51cq7gucp). See [this doc](https://docs.google.com/document/d/1TLRIOurdRK1HE7nDPqk6d4zhfjpgchBd9tU6PQpbwxw/edit?usp=sharing)

This PR will:
- Add driver annotation, driver tiers and mutation status filtering to _/filtered-samples/fetch_, _/mutated-genes/fetch_, _/fusion-genes/fetch_, _/cna-genes/fetch_, _/alteration-enrichments/fetch_, _/copy-number-enrichments/fetch_ and _/mutation-enrichments/fetch_ endpoints.
- Add database indexes for different fields used for queries.

### CODE REVIEW

File: persistence/persistence-api/src/main/java/org/cbioportal/persistence/MolecularDataRepository.java [hunk 1]
```diff
@@ -3,11 +3,8 @@
 import java.util.List;
 import java.util.Map;
 
-import org.cbioportal.model.GeneMolecularAlteration;
-import org.cbioportal.model.GenericAssayMolecularAlteration;
-import org.cbioportal.model.GenesetMolecularAlteration;
+import org.cbioportal.model.*;
```
Comment (Luke-Sikina) [inline persistence/persistence-api/src/main/java/org/cbioportal/persistence/MolecularDataRepository.java:6]: Your only changes in here are wildcarding an import and a whitespace change. I don't think this should be included in the PR.

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/AlterationCountsMapper.xml [hunk 2]
```diff
@@ -20,10 +20,11 @@
             FROM mutation_event
             INNER JOIN mutation ON mutation_event.MUTATION_EVENT_ID = mutation.MUTATION_EVENT_ID
             INNER JOIN gene ON mutation_event.ENTREZ_GENE_ID = gene.ENTREZ_GENE_ID
+            <include refid="fromIncludeCustomAnnotationsMutation"/>
             <where>
                 <choose>
                     <when test="mutationTypes.hasNone()">NULL</when>
-                    <when test="!mutationTypes.hasAll()">
+                    <when test="not mutationTypes.hasAll()">
```
Comment (Luke-Sikina) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/AlterationCountsMapper.xml:27]: Are these the same? Why do we prefer one over the other?

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/AlterationCountsMapper.xml [hunk 3]
```diff
@@ -149,7 +159,7 @@
     </select>
 
     <select id="getSampleInternalIds" resultType="Integer">
-        SELECT sample.INTERNAL_ID
+        SELECT DISTINCT sample.INTERNAL_ID
```
Comment (Luke-Sikina) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/AlterationCountsMapper.xml:162]: Why did you need to add DISTINCT? I looked briefly, but I couldn't figure out how you were seeing duplicates.

**Review by Luke-Sikina — COMMENTED** [review]
Will review more tomorrow. Looks pretty reasonable so far. I'd start in the StudyViewFilter and it's applier - that's where I had the most significant feedback.

File: service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java [hunk 4]
```diff
@@ -219,6 +222,43 @@ public Integer getNumberOfSamplesInMolecularProfile(String molecularProfileId) {
...
+        List<Integer> entrezGeneIds = geneQueries.stream().map(q -> q.getEntrezGeneId()).collect(Collectors.toList());
+        List<GeneMolecularData> molecularDataList = getMolecularDataInMultipleMolecularProfiles(molecularProfileIds, sampleIds, entrezGeneIds, projection);
+
+        // All CNA events that match requested geneQueries
+        List<DiscreteCopyNumberData> copyNumberData = discreteCopyNumberRepository
+            .getDiscreteCopyNumbersInMultipleMolecularProfilesByGeneQueries(molecularProfileIds, sampleIds,
+                geneQueries, projection);
+        
+        // molecularProfile->entrezGeneId->sampleId->alterationType lookup table for CNA events
+        Map<String, Map<Integer, Map<String, Map<Integer, String>>>> cnaEventLookup = copyNumberData.stream()
```
Comment (kalletlak) [inline service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java:240]: instead of multi-grouping why not create multikey map?

File: web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java [hunk 5]
```diff
@@ -66,11 +66,27 @@
         Select<CNA> cnaEventTypes = allOptionsSelected(alterationEventTypes.getCopyNumberAlterationEventTypes()) ?
             Select.all() : Select.byValues(selectedCnas);
 
+        Select<String> selectedTiers = Select.byValues(
+            alterationEventTypes.getSelectedTiers().entrySet().stream()
+                .filter(e -> e.getValue())
+                .map(e -> e.getKey()));
+        if (alterationEventTypes.getSelectedTiers().keySet().size() > 0
```
Comment (kalletlak) [inline web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java:73]: I have a question here. Consider there are 3 tiers - tier1, tier2, tier3. If alterationEventTypes.selectedTiers is `{"tier1" : true, "tier2": true}` then is this condition true and selectedTiers is said to have all values?

File: web/src/main/java/org/cbioportal/web/StudyViewController.java [hunk 6]
```diff
@@ -338,7 +346,15 @@
...
-                Select.all());
+                Select.all(),
+                interceptedStudyViewFilter.isIncludeDriver(),
+                interceptedStudyViewFilter.isIncludeVUS(),
+                interceptedStudyViewFilter.isIncludeUnknownOncogenicity(),
+                selectedTiers,
+                interceptedStudyViewFilter.isIncludeUnknownTier(),
+                interceptedStudyViewFilter.isIncludeGermline(),
+                interceptedStudyViewFilter.isIncludeSomatic(),
+                interceptedStudyViewFilter.isIncludeUnknownStatus());
```
[the remark this answers is not in the document]
Comment (kalletlak) [reply web/src/main/java/org/cbioportal/web/StudyViewController.java:357]: agreed, GeneFilterQuery.java have similar fields. may be extract them in a new class

File: service/src/main/java/org/cbioportal/service/impl/AlterationDriverAnnotationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    private final Set<String> NO_DATA_TIERS = new HashSet<>(Arrays.asList(null, "", "NA"));
+    
+    public CustomDriverAnnotationReport getCustomDriverAnnotationProps(List<String> molecularProfileIds) {
+        
+        List<AlterationDriverAnnotation> rows = alterationDriverAnnotationRepository
+            .getAlterationDriverAnnotations(molecularProfileIds);
+        
+        Set<String> tiers = rows.stream()
+            .filter(d -> !NO_DATA_TIERS.contains(d.getDriverTiersFilter()))
+            .map(AlterationDriverAnnotation::getDriverTiersFilter)
```
Comment (Luke-Sikina) [inline service/src/main/java/org/cbioportal/service/impl/AlterationDriverAnnotationServiceImpl.java:27]: Put the map first so you only have to call the getter once
```suggestion
            .map(AlterationDriverAnnotation::getDriverTiersFilter)
            .filter(driverTiersFilter -> !NO_DATA_TIERS.contains(driverTiersFilter))
```

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml [hunk 8]
```diff
@@ -2,7 +2,6 @@
 <!DOCTYPE mapper PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN" "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
 
 <mapper namespace="org.cbioportal.persistence.mybatis.MolecularDataMapper">
```
Comment (Luke-Sikina) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml:4]: Only whitespace changes in this file. Remove from PR.

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularProfileMapper.xml [hunk 9]
```diff
@@ -145,5 +145,4 @@
                          genetic_profile.STABLE_ID = #{referredMolecularProfileId}
                        )
     </select>
-
```
Comment (Luke-Sikina) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularProfileMapper.xml:148]: Only whitespace changes in this file. Remove from PR.

File: service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java [hunk 10]
```diff
@@ -219,6 +222,43 @@ public Integer getNumberOfSamplesInMolecularProfile(String molecularProfileId) {
         return molecularDataList;
     }
 
+    @Override
+    public List<GeneMolecularData> getMolecularDataInMultipleMolecularProfilesByGeneQueries(List<String> molecularProfileIds,
+                                                                                            List<String> sampleIds,
+                                                                                            List<GeneFilterQuery> geneQueries,
+                                                                                            String projection) {
+        // Molecular alterations for all genes in the geneQueries
+        List<Integer> entrezGeneIds = geneQueries.stream().map(q -> q.getEntrezGeneId()).collect(Collectors.toList());
```
Comment (Luke-Sikina) [inline service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java:231]: In general, I prefer method references, as they give a rough idea of the type at that point in the stream. Did you just avoid using one here to stay under 120 chars?
```suggestion
        List<Integer> entrezGeneIds = geneQueries.stream().map(GeneFilterQuery::getEntrezGeneId).collect(Collectors.toList());
```

**Review by Luke-Sikina — COMMENTED** [review]
Done reviewing!

File: web/src/main/java/org/cbioportal/web/parameter/StudyViewFilter.java [hunk 11]
```diff
@@ -29,6 +31,78 @@
     private List<GenericAssayDataFilter> genericAssayDataFilters;
     private List<List<String>> caseLists;
     private List<ClinicalDataFilter> customDataFilters;
+    private Map<String,Boolean> selectedTiers;
+    private boolean includeDriver;
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/parameter/StudyViewFilter.java:35]: Ok, all parameters were moved to a new web parameter class _AlterationFilter_ which is a member of _StudyViewFilter_.

File: web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java [hunk 12]
```diff
@@ -294,27 +294,41 @@ public SampleIdentifier apply(Sample sample) {
...
 
-            for (List<SingleGeneQuery> geneQueries : genefilter.getSingleGeneQueries()) {
+            for (List<GeneFilterQuery> geneQueries : genefilter.getGeneQueries()) {
                 List<String> studyIds = new ArrayList<>();
                 List<String> sampleIds = new ArrayList<>();
 
-                List<String> hugoGeneSymbols = geneQueries.stream().map(SingleGeneQuery::getHugoGeneSymbol)
-                        .collect(Collectors.toList());
+                List<String> hugoGeneSymbols = geneQueries.stream()
+                    .map(org.cbioportal.model.GeneFilterQuery::getHugoGeneSymbol).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java:302]: Fixed!

File: web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java [hunk 13]
```diff
@@ -294,27 +294,41 @@ public SampleIdentifier apply(Sample sample) {
...
+
+                Map<String, Integer> symbolToEntrezGeneId = geneService
+                    .fetchGenes(new ArrayList<>(hugoGeneSymbols),
+                        GeneIdType.HUGO_GENE_SYMBOL.name(), Projection.SUMMARY.name())
+                    .stream().collect(Collectors.toMap(x -> x.getHugoGeneSymbol(), x -> x.getEntrezGeneId()));
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java:307]: Ah, I did not know. Updated!

File: web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java [hunk 14]
```diff
@@ -294,27 +294,41 @@ public SampleIdentifier apply(Sample sample) {
...
-                List<Integer> entrezGeneIds = geneService
-                        .fetchGenes(hugoGeneSymbols, GeneIdType.HUGO_GENE_SYMBOL.name(), Projection.SUMMARY.name())
-                        .stream().map(gene -> gene.getEntrezGeneId()).collect(Collectors.toList());
+                geneQueries.removeIf(
+                    q -> !symbolToEntrezGeneId.containsKey(q.getHugoGeneSymbol())
+                );
 
+                geneQueries.stream().forEach(
+                    q -> q.setEntrezGeneId(symbolToEntrezGeneId.get(q.getHugoGeneSymbol()))
+                );
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java:315]: Well, I do not quite remember why I added this. It appears that this new logic is there to handle the event where the _geneService_ returns EntrezGeneIds for less genes (because of incomplete mapping) for the number of GeneQueries that are passed in. Normally the EntrezGeneIds were passed to the _getMutationsInMultipleMolecularProfiles_ method, effectively reducing the search space to less genes. In my PR it is the list of GeneQueries that is passed to _getMutationsInMultipleMolecularProfiles_. So, I first filter out the queries that do not relate to correctly mapped genes

Does that make sense?.

File: web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java [hunk 15]
```diff
@@ -377,27 +398,36 @@ public SampleIdentifier apply(Sample sample) {
...
+                                    GeneIdType.HUGO_GENE_SYMBOL.name(), Projection.SUMMARY.name())
+                                .stream().collect(Collectors.toMap(x -> x.getHugoGeneSymbol(), x -> x.getEntrezGeneId()));
+
+                            filteredGeneQueries.removeIf(
+                                q -> !symbolToEntrezGeneId.containsKey(q.getHugoGeneSymbol())
+                            );
+
+                            filteredGeneQueries.stream().forEach(
+                                q -> q.setEntrezGeneId(symbolToEntrezGeneId.get(q.getHugoGeneSymbol()))
+                            );
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java:421]: I see your point. However, I merely updated the logic of the existing methods in _StudyViewFilterApplier_ that is present in master branch. I propose to file an issue on GitHub for this, but to consider your proposal as out of scope for this PR. Do you agree? If so, I will create the issue.

File: web/src/main/java/org/cbioportal/web/StudyViewController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/StudyViewController.java:357]: Ok, done.

File: persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/MolecularDataMyBatisRepository.java [hunk 16]
```diff
@@ -1,9 +1,6 @@
 package org.cbioportal.persistence.mybatis;
 
-import org.cbioportal.model.GeneMolecularAlteration;
-import org.cbioportal.model.GenericAssayMolecularAlteration;
-import org.cbioportal.model.GenesetMolecularAlteration;
-import org.cbioportal.model.MolecularProfileSamples;
+import org.cbioportal.model.*;
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/MolecularDataMyBatisRepository.java:3]: Done!

File: persistence/persistence-api/src/main/java/org/cbioportal/persistence/MolecularDataRepository.java (hunk 1 above)
Comment (pvannierop) [reply persistence/persistence-api/src/main/java/org/cbioportal/persistence/MolecularDataRepository.java:6]: Done!

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/AlterationCountsMapper.xml (hunk 2 above)
Comment (pvannierop) [reply persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/AlterationCountsMapper.xml:27]: They are the same in meaning. I switched to `not` because in some instances the evaluation of `!` caused problems. I guess OGNL (the expression language used by MyBatis) can be a bit unpredictable in some evaluations.

File: model/src/main/java/org/cbioportal/model/AnnotationSourceFilter.java [hunk 17]
```diff
@@ -0,0 +1,25 @@
+package org.cbioportal.model;
+
+import org.cbioportal.model.util.Select;
+
+public class AnnotationSourceFilter {
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply model/src/main/java/org/cbioportal/model/AnnotationSourceFilter.java:5]: Ah it is a relic of the past. Deleted.

File: model/src/main/java/org/cbioportal/model/CopyNumberAlterationEventType.java [hunk 18]
```diff
@@ -0,0 +1,17 @@
+package org.cbioportal.model;
+
+public enum CopyNumberAlterationEventType {
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply model/src/main/java/org/cbioportal/model/CopyNumberAlterationEventType.java:3]: Indeed, corrected via a rebase to master (I removed this class in a previous PR)

File: model/src/main/java/org/cbioportal/model/GeneFilterQuery.java [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+    private Integer entrezGeneId;
+    private List<CNA> alterations;
+    private boolean includeDriver;
+    private boolean includeVUS;
+    private boolean includeUnknownOncogenicity;
+    private boolean includeGermline;
+    private boolean includeSomatic;
+    private boolean includeUnknownStatus;
+    private Map<String, Boolean> selectedTiers;
+    @JsonIgnore
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply model/src/main/java/org/cbioportal/model/GeneFilterQuery.java:22]: Internally selected tiers are represented by the Select class that can have stated _all_, _none_ and _some_. The API exposes selected tiers as a {[tier:string] : boolean} object. The boolean state of this object is used to evaluate the _all_ or _none_ state of the Select instance that represents the selected tiers. The internal representation is not exposed via the API.

File: model/src/main/java/org/cbioportal/model/util/QueryElement.java [hunk 20]
```diff
@@ -0,0 +1,5 @@
+package org.cbioportal.model.util;
+
+public enum QueryElement {
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply model/src/main/java/org/cbioportal/model/util/QueryElement.java:3]: What do you think of this?

```
// QueryElement represents three types of behavior for a SQL statement in MyBatisMappers
// INACTIVE: exclude Query element from results
// ACTIVE: include Query element in results
// PASS: do not apply on Query element
// Note: the specific behavior the represented by the three states of
// QueryElement is determined by the mapper-xml.
```

[the remark this answers is not in the document]
Comment (pvannierop) [reply model/src/main/java/org/cbioportal/model/util/QueryElement.java:3]: What do you think of this?

```
// QueryElement represents three types of behavior for a SQL statement in MyBatisMappers
// INACTIVE: exclude Query element from results
// ACTIVE: include Query element in results
// PASS: do not apply on Query element
// Note: the specific behavior the represented by the three states of
// QueryElement is determined by the mapper-xml.
```

File: persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/AlterationMyBatisRepository.java [hunk 21]
```diff
@@ -70,16 +96,29 @@
...
+                                                          Select<CNA> cnaEventTypes,
+                                                          boolean includeDriver,
+                                                          boolean includeVUS,
+                                                          boolean includeUnknownOncogenicity,
+                                                          Select<String> selectedTiers, boolean includeUnknownTier) {
 
         if (molecularProfileCaseIdentifiers == null || molecularProfileCaseIdentifiers.isEmpty()
-            || cnaEventTypes == null || cnaEventTypes.hasNone()) {
+            || cnaEventTypes == null || cnaEventTypes.hasNone()
+            || (!includeDriver && !includeVUS && !includeUnknownOncogenicity && selectedTiers.hasNone() && !includeUnknownTier)) {
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/AlterationMyBatisRepository.java:121]: Ok, I updated the code. But I do not think it _looks_ better. I must say is it more understandable about what is tested though. What do you think?

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/DiscreteCopyNumberMapper.xml [hunk 22]
```diff
@@ -37,82 +36,118 @@
...
-                </foreach>
-            </if>
-            <if test="alterationTypes != null and !alterationTypes.isEmpty()">
-                AND cna_event.ALTERATION IN
-                <foreach item="item" collection="alterationTypes" open="(" separator="," close=")">
-                    #{item}
-                </foreach>
-            </if>
-        </where>
+        genetic_profile.STABLE_ID = #{molecularProfileId}
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/DiscreteCopyNumberMapper.xml:39]: You are correct. I changed this so all _where_ sql fragments not contain `<where>` tags themselves to that they can be combined like for instance so (imaginary example):

```
        <where>
            <include refid="whereBySampleListId"/>
            <include refid="whereGenes"/>
        </where>
```

I personally think this is better practice. I can revert if you want. Just let met know.

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MutationMapper.xml [hunk 23]
```diff
@@ -64,6 +64,9 @@
                 mutation.GENETIC_PROFILE_ID = alteration_driver_annotation.GENETIC_PROFILE_ID
             and mutation.SAMPLE_ID = alteration_driver_annotation.SAMPLE_ID
             and mutation.MUTATION_EVENT_ID = alteration_driver_annotation.ALTERATION_EVENT_ID
+        <if test="_parameter.containsKey('projection') and (projection == 'SUMMARY' or projection == 'DETAILED')">
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MutationMapper.xml:67]: Yes, depending on the calling method.

File: web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (Luke-Sikina) [reply web/src/main/java/org/cbioportal/web/util/StudyViewFilterApplier.java:315]: Makes sense, thanks for explaining.

File: model/src/main/java/org/cbioportal/model/util/QueryElement.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (kalletlak) [reply model/src/main/java/org/cbioportal/model/util/QueryElement.java:3]: looks good. thanks!

File: service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java [hunk 24]
```diff
@@ -87,28 +88,27 @@ public BaseMeta getMetaDiscreteCopyNumbersInMolecularProfileBySampleListId(Strin
...
-    public List<DiscreteCopyNumberData> getDiscreteCopyNumbersInMultipleMolecularProfiles(List<String> molecularProfileIds, 
-                                                                                          List<String> sampleIds, 
+    public List<DiscreteCopyNumberData> getDiscreteCopyNumbersInMultipleMolecularProfiles(List<String> molecularProfileIds,
+                                                                                          List<String> sampleIds,
                                                                                           List<Integer> entrezGeneIds,
-                                                                                          List<Integer> alterationTypes, 
+                                                                                          List<Integer> alterationTypes,
                                                                                           String projection) {
-        
-        if (isHomdelOrAmpOnly(alterationTypes)) {
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply service/src/main/java/org/cbioportal/service/impl/DiscreteCopyNumberServiceImpl.java:96]: Hmm that is such a long time ago. I have restored this functionality in the _getDiscreteCopyNumbersInMultipleMolecularProfiles_ and in the _getDiscreteCopyNumbersInMultipleMolecularProfilesByGeneQueries_ method.

File: service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java (hunk 4 above)
Comment (pvannierop) [reply service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java:240]: Yes, much nicer. Thanks for this suggestion. So updated...

File: web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java (hunk 5 above)
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java:73]: That is correct!

File: service/src/main/java/org/cbioportal/service/impl/AlterationDriverAnnotationServiceImpl.java (hunk 7 above)
Comment (pvannierop) [reply service/src/main/java/org/cbioportal/service/impl/AlterationDriverAnnotationServiceImpl.java:27]: Ah yes, thanks for that suggestion. I never look at it that way. I'll keep that in mind in the future. So, updated!

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml (hunk 8 above)
Comment (pvannierop) [reply persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularDataMapper.xml:4]: Done!

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularProfileMapper.xml (hunk 9 above)
Comment (pvannierop) [reply persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/MolecularProfileMapper.xml:148]: Done!

File: service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java (hunk 10 above)
Comment (pvannierop) [reply service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java:231]: No that was not a consideration at all. I have accepted you code improvement.

Comment (pvannierop) [reply service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java:231]: Yes, updated!

File: service/src/main/java/org/cbioportal/service/AlterationCountService.java [hunk 25]
```diff
@@ -14,64 +14,126 @@
                                                           boolean includeMissingAlterationsFromGenePanel,
                                                           Select<MutationEventType> mutationEventTypes,
                                                           Select<CNA> cnaEventTypes,
-                                                          QueryElement searchFusions);
+                                                          QueryElement searchFusions,
```
Comment (Luke-Sikina) [inline service/src/main/java/org/cbioportal/service/AlterationCountService.java:17]: Can you use the `AlterationFilter` object that you made rather than passing all these fields? This comment applies for all services, repositories, and mappers that have this pattern.

**Review by Luke-Sikina — COMMENTED** [review]
Thank you for the changes, only one remaining suggestion.

File: web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java (hunk 5 above)
Comment (kalletlak) [reply web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java:73]: isn't that a bug then?

File: persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/AlterationMyBatisRepository.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (kalletlak) [reply persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/AlterationMyBatisRepository.java]: looks good to me

File: web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java (hunk 5 above)
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java:73]: How do you mean?

Comment (kalletlak) [reply web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java:73]: this condition is will be true for both
`{"tier1" : true, "tier2": true, "tier3" : true}` and `{"tier1" : true, "tier2": true}` even if there is no information about `tier3`
Does this means the default value is true if tier is not present in request object?

File: service/src/main/java/org/cbioportal/service/AlterationCountService.java (hunk 25 above)
Comment (pvannierop) [reply service/src/main/java/org/cbioportal/service/AlterationCountService.java:17]: I have collapsed all filter options in a single AlterationFilter object. This was a quite substantial refactoring. A lot of files changes. Can you re-review this?

File: service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java [hunk 26]
```diff
@@ -219,6 +222,34 @@ public Integer getNumberOfSamplesInMolecularProfile(String molecularProfileId) {
...
+                                                                                            List<GeneFilterQuery> geneQueries,
+                                                                                            String projection) {
+        // Molecular alterations for all genes in the geneQueries
+        List<Integer> entrezGeneIds = geneQueries.stream()
+            .map(GeneFilterQuery::getEntrezGeneId)
+            .collect(Collectors.toList());
+        List<GeneMolecularData> molecularDataList = getMolecularDataInMultipleMolecularProfiles(molecularProfileIds, sampleIds, entrezGeneIds, projection);
+
+        // All CNA events that match requested geneQueries
+        List<DiscreteCopyNumberData> copyNumberData = discreteCopyNumberRepository
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java:237]: I do not spot the cyclic invocation. Can you be more specific?

File: web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java (hunk 5 above)
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java:73]: Ah I understand now. No, it evaluates whether to search for all types based on all keys being true. So this logic expects the frontend to send over a complete map of all options. If the frontend does not do this, there wil be a problem indeed. There is no other way for us to evaluate whether all options are selected I can see. Shall we discuss this with @alisman to see whether this a correct approach?

**Review by Luke-Sikina — APPROVED** [review]
You have addressed all my comments. Thank you!

File: web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java (hunk 5 above)
Comment (kalletlak) [reply web/src/main/java/org/cbioportal/web/AlterationEnrichmentController.java]: This shouldn't be an issue if the request is coming for frontend. But if the user is trying to query API directly then it might be an issue.

File: service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (kalletlak) [reply service/src/main/java/org/cbioportal/service/impl/MolecularDataServiceImpl.java:237]: Sorry, confused DiscreteCopyNumberService with DiscreteCopyNumberRepository

**Review by kalletlak — APPROVED** [review]
LGTM except one thing, SelectedTiers object.

**Comment by onursumer** [comment]
Maybe we can break this down some more before estimating:

- [ ] set up test instance
- [ ] create release plan
- [ ] remaining coding effort
- [ ] code review effort

**Comment by inodb** [comment]
Superseded by #8647
